### ISSUE TITLE
Performance comparison between source table and AMT + MV.

### ISSUE THREAD

**Issue by moweonlee** [ClickHouse/ClickHouse#16457]
I'd like to ask a question in the result of my performance test.

I have made several MV+AMT tables to reduce rows efficiently when hourly queries and daily queries are executed.

I'm  basically getting data for every 10 seconds. So executing hourly/daily query on this source table is not efficient in I/O side especially. 

## Backgrounds 

First I made a source table like below.
Data in inserted into source table every 10 seconds. 

![image](https://user-images.githubusercontent.com/16459517/97391918-b8359080-1923-11eb-89b0-8645987699d8.png)

And an AMT + MV table is added to aggregate data hourly. 

With "SimpleAggregateFunction( groupUniqArrayArray  , Array(String) ))", I was able to replace "uniqState/uniqMerge" and see aggregate result directly..

![image](https://user-images.githubusercontent.com/16459517/97392054-1a8e9100-1924-11eb-905e-45f179b9eaa9.png)

In addition to hour, another AMT+ MV table to aggregate data for daily query.

![image](https://user-images.githubusercontent.com/16459517/97392130-5de8ff80-1924-11eb-8086-b5f14ce1c973.png)

Rows count are like this. ( tested with 20% of data )
And it is after Table Optimization with "OPTIMEZE TABLE {3tables} FINAL"

- Source table : 173645997
- Hourly MV : 11788335  ( 93% reduction ) 
- Daily MV : 1781238 ( 99% reduction ) 

## Question

To get the sum of page views and number of unique users in each table, test queries are executed 20 times for each.

![image](https://user-images.githubusercontent.com/16459517/97395386-3431d780-1928-11eb-9a2d-5c34e81ac043.png)

Even 93% and 99% reductions in hourly and daily AMT, query performances are....

- slower than source table in case of hourly table.
- slightly faster than source table in case of daily table. 

**Can someone please explain the reason of this slow performance  ?**

**And any of your advises and opinion are welcome.**

**Basically what I want to do is reducing data in AMT to get better performance."

## Test Queries 

- source table query
```
select os, round( sum( divide( pv, length(users) ) ) ) as pv , uniq(arrayJoin( users) ) as uv from dlcs_stms_array_basic_20010101 group by os order by os asc;

```

- hourly mv query
```
select os, round( sum( divide( pv, length(user) ) ) ) as pv , uniq(arrayJoin( user) ) as uv from damtlcs_stms_array_basic_hourly_20010101 group by os order by os asc;

```

- daily mv query

```
select os, round( sum( divide( pv, length(user) ) ) ) as pv , uniq(arrayJoin( user) ) as uv from damtlcs_stms_array_basic_daily_20010101 group by os order by os asc;

```

**Comment by den-crane** [comment]
Third time I am looking at this issue and I don't want retype pictures. Why did you post pictures?

**Comment by den-crane** [comment]
I guess the problem is in `uniq(arrayJoin( user) )` and can be solved without arrayJoin.

**Comment by moweonlee** [comment]
Oh I guess arrayJoin( users ) still need similar workload with source table because uniq() still need to process every row like what uniq() do in the source table .

But can it be said that uniq(arrayJoin( users ) ) need less memory than uniq(users) in source table which doesn't me merged at all. ?

**Comment by moweonlee** [comment]
> Third time I am looking at this issue and I don't want to retype pictures. Why did you post pictures?

Oh I guess picture would be more easy to understand.

I promise there will be paste-able text from next time.

**Comment by den-crane** [comment]
No steps to reproduce the issue.
